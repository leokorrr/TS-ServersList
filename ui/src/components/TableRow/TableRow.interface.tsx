export default interface TableRowInterface {
    name: string;
    status: 'ONLINE' | 'OFFLINE' | 'REBOOTING';
    _id: number;
}