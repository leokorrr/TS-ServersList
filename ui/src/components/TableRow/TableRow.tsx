import React from 'react'
import TableRowInterface from './TableRow.interface'
import more from '../../images/more.svg'
import serverOn from '../../images/serverOn.svg'
import serverOff from '../../images/serverOff.svg'
import useTableRow from '../../hooks/useTableRow'

const TableRow: React.FC<TableRowInterface> = ({_id, name, status}) => {
    const {
        serverStatus,
        showDropdown,
        handleTurnOff,
        handleTurnOn,
        handleChangeDropdownVisibility,
        handleReboot,
        wrapperRef
    } = useTableRow(status)
    return (
        <tr className="table-row">
            <td className="table-row__cell table-row__cell--name">{name}</td>
            <td className="table-row__cell table-row__cell--status">
                {(serverStatus === 'ONLINE') && <div className="green"><img src={serverOn} alt="on"/>{serverStatus}</div>}
                {(serverStatus === 'OFFLINE') && <div><img src={serverOff} alt="off"/>{serverStatus}</div>}
                {(serverStatus === 'REBOOTING') && <div className="grey">{serverStatus}...</div>}
            </td>
            <td className="table-row__cell">
                <div className="table-row__cell-more">
                    <div>
                        <button className="table-row__button-more" onClick={() => handleChangeDropdownVisibility(true)}><img src={more} alt="more"/></button>
                        {showDropdown && serverStatus !== 'REBOOTING' &&
                            <ul className="table-row__dropdown" ref={wrapperRef}>
                                <li>
                                    {(serverStatus === 'ONLINE') ?  
                                        <button className="table-row__dropdown-btn" onClick={() => handleTurnOff(_id)}>Turn off</button> 
                                        : 
                                        <button className="table-row__dropdown-btn"  onClick={() => handleTurnOn(_id)}>Turn on</button>}
                                </li>
                                <li>
                                    {(serverStatus === 'ONLINE') && 
                                        <button className="table-row__dropdown-btn" 
                                            onClick={() => handleReboot(_id)}>Reboot</button>}
                                </li>
                            </ul> 
                        }
                        {showDropdown && serverStatus === 'REBOOTING' && 
                            <ul className="table-row__dropdown" ref={wrapperRef}>
                                <li>
                                    <button disabled className="table-row__dropdown-btn table-row__dropdown-btn--disabled">Rebooting...</button>
                                </li>
                            </ul>}
                    </div>
                </div>
            </td>
        </tr>
    )
}
    
export default TableRow