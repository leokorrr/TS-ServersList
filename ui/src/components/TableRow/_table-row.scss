.table-row {
    &__dropdown {
        padding: 0;
        list-style: none;
        background-color: $color-white;
        position: absolute;
        top: 0;
        right: -0.5rem;
        box-shadow: 0 0 0.7rem $color-shadow;
        width: 15rem;
        z-index: 2;

        li {
            width: 100%;
        }
    }

    &__dropdown-btn {
        width: 100%;
        height: 100%;
        background-color: $color-white;
        border: none;
        text-align: left;
        padding: 1.5rem 2rem;
        transition: .2s;

        &:hover {
            cursor: pointer;
            background-color: darken($color-white, 5%);
        }

        &:focus {
            outline: none;
        }

        &--disabled {
            &:hover {
                cursor:not-allowed;
                background-color: $color-white;
            }
        }
    }

    &__button-more {
        border: none;
        background-color: transparent;

        &:hover {
            cursor: pointer;
        }

        &:focus {
            outline: none;
        }
    }

    &__cell-more {
        & > div {
            display: flex;
            flex-direction: row-reverse;
        }
    }

    &__cell {
        border-bottom: 0.2rem solid $color-table-border;
        padding: 2.2rem 4.2rem;
        text-align: left;
        font-size: 1.4rem;
        font-weight: 600;
        position: relative;

        &:first-of-type {
            width: 27.8rem;
        }

        &--header {
            padding-top: 4.1rem;
            color: $color-table-grey;
            text-transform: uppercase;
        }

        &--status {
            & > div {
                display: flex;
                align-items: center;
            }

            img {
                margin-right: 0.8rem;
            }

            .green {
                img {
                    height: 0.7rem;
                    width: 0.7rem;
                }
            }

            .red {
                img {
                    height: 1.1rem;
                    width: 1rem;
                }
            }
        }
    }
}