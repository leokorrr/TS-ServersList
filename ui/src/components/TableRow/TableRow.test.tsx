import React from 'react'
import TableRow from './TableRow'
import { shallow, configure } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

const _id = 1
const name = ''
const status = 'ONLINE'
configure({ adapter: new Adapter() })
it("Table Header rendered without crashing", () => {
    shallow(<TableRow _id={_id} name={name} status={status}/>);
})
it("should render initial layout", () => {
    const component = shallow(<TableRow _id={_id} name={name} status={status}/>);
    expect(component.getElements()).toMatchSnapshot();
})