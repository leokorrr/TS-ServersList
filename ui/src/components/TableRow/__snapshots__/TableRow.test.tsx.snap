// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render initial layout 1`] = `
Array [
  <tr
    className="table-row"
  >
    <td
      className="table-row__cell table-row__cell--name"
    >
      
    </td>
    <td
      className="table-row__cell table-row__cell--status"
    >
      <div
        className="green"
      >
        <img
          alt="on"
          src="serverOn.svg"
        />
        ONLINE
      </div>
    </td>
    <td
      className="table-row__cell"
    >
      <div
        className="table-row__cell-more"
      >
        <div>
          <button
            className="table-row__button-more"
            onClick={[Function]}
          >
            <img
              alt="more"
              src="more.svg"
            />
          </button>
        </div>
      </div>
    </td>
  </tr>,
]
`;
